src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -488,49 +501,41 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
             # We don't flatten when there are duplicate columns.
             can_be_flattened = False
             disable_next_level_flatten = True
-        elif self.flatten_disabled or self.has_clause_using_columns:
+        elif self.flatten_disabled:
+            can_be_flattened = False
+        elif (
+            self.where
+            and self.where.dependent_columns
+            and any(
+                new_column_states[_col].change_state in (ColumnChangeState.NEW,)
+                for _col in self.where.dependent_columns.intersection(
+                    new_column_states.active_columns
+                )
+            )
+        ):
+            can_be_flattened = False
+        elif self.order_by:  # TODO: implement orderby
```

Comment:
Not sure about the policy in open source repo but might be good to have a JIRA?

Reply:
SNOW-704094 covers both filter and orderby, I just want to separate the PR's for easier review.